### PR TITLE
remote-week2: initial commit

### PR DESCRIPTION
**Pull request by yingyinglininin**
IP address can access to my APIs: 13.215.241.188 
- app.js: port 4000
- adminAPI.js: port 3000

Database name: http://database-1.c8of2lxxr8vl.ap-southeast-1.rds.amazonaws.com/
Change summary:
- Create assignment.user table in MySQL
- Add app.js
- Add adminAPI.js

Understanding Check:
1. I have created a user table according to the specifications outlined in assignment 2. I've designated the 'id' column as the primary key and set the other columns with appropriate data types and default values. Here's the schema for the user table: image image
2. Status code is standardized three-digit numbers returned by an API in response to a client’s request to indicate the outcome of a specific HTTP request. The common HTTP status codes include: 200 (OK), 403 (Forbidden), 404 (Not Found). 
The following is the HTTP Status Code part of Restful API document: https://restfulapi.net/http-status-codes/ 
It is not mandatory for all APIs, but it’s highly recommended to use because it enhances the usability and reliability of our APIs. We can quickly know the status of our client requests and determine the nature of the error.
3. After developing our APIs, we can use some tools such as Postman (test directly in GUI) and Swagger (Write Swagger document) to test our APIs. We can start our backend service in localhost or deploy it on cloud servers such as Amazon EC2. Next, use the client (API testing tools) to send the request with an API URL. Finally, we can get the status code and the response data from the server.
4. Firstly, we can enter EC2 with an ssh connection. Next, we can use commands such as `screen`, `tmux` and `nohup` to keep remote SSH sessions and processes running after disconnection.
image
image
image

5. The following is the interation diagram between client, firewall, EC2 instance and RDS database: image

6. To allow clients to access APIs without specifying a port number in the URL, we can use standard ports. That is, web applications typically run on HTTP (port 80) or HTTPS (port 443) by default. By using these standard ports, we can omit the port number from the URL. However, we will encounter an error message "permission denied 0.0.0.0:443" because we need root permissions to access port < 1024. Thus, using ngnix as a reverse proxy server may be a solution. (To be researched) image
- Stack Overflow: https://stackoverflow.com/questions/50112186/error-listen-eacces-0-0-0-0443
- Setup Nginx as reverse proxy on EC2 instance: https://medium.com/@imageadhikari/setup-nginx-as-reverse-proxy-on-ec2-instance-3c6820b6467b
7. A web server is a software component that delivers static data such as images, files, and in response to client requests. An application server adds business logic to compute the web server's response. In summary, a web server only hosts websites and delivers responses to simple requests. On the other hand, application servers handle business logic to generate dynamic content by connecting with enterprise systems, services and databases.

### CODE REVIEW

File: Week2/adminAPI.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// get the client
+const mysql = require('mysql2');
+const config = require('config');
+const express = require('express');
+const bodyParser = require('body-parser');
```
Comment (addie-tyc) [inline Week2/adminAPI.js]: 現在用的 express 版本還有需要引入 body-parser 嗎？

File: Week2/adminAPI.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            res.status(409).json({ error: 'Email Already Exists' });
+          } else {
+              const responseData = {
+                "data": {
+                  "user": {
+                    "id": result.insertId,
+                    "name": user.name,
+                    "email": user.email
+                  },
+                  "request-date": new Date().toUTCString()
```
Comment (addie-tyc) [inline Week2/adminAPI.js]: request-date 是要直接從 request header 拿對應的 key 喔

File: Week2/adminAPI.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        } else {
+          const user = results[0];
+          const responseData = {
+            "data": {
+              "user": {
+                "id": user.id,
+                "name": user.name,
+                "email": user.email
+              },
+              "request-date": new Date().toUTCString()
```
Comment (addie-tyc) [inline Week2/adminAPI.js]: 同上

**Review by addie-tyc — CHANGES_REQUESTED** [review]
understanding check 很完整～小地方需修正

**Comment by addie-tyc** [comment]
healthcheck 好像沒有對外開放？

**Review by addie-tyc — CHANGES_REQUESTED** [review]
healthcheck OK，但 user 好像壞了？
截圖 2023-10-13 21 31 47

File: Week2/adminAPI.js (hunk 1 above)
Comment (yingyinglininin) [reply Week2/adminAPI.js]: 因為之前在嘗試 `app.use(body-parser.json());` 的語法，以為需要導入 body-parser 才可以，不過研究後發現，現在的版本直接使用`app.use(express.json());` 即可

File: Week2/adminAPI.js (hunk 2 above)
Comment (yingyinglininin) [reply Week2/adminAPI.js]: Done

File: Week2/adminAPI.js (hunk 3 above)
Comment (yingyinglininin) [reply Week2/adminAPI.js]: Done

**Comment by yingyinglininin** [comment]
> healthcheck 好像沒有對外開放？

先前因為沒有在 EC2 開啟對應的 port 4000 權限，所以沒有對外開放，目前已經順利開啟！

**Comment by yingyinglininin** [comment]
> healthcheck OK，但 user 好像壞了？ 截圖 2023-10-13 21 31 47

剛剛發現是因為少了` app.use(express.json()); ` 所以導致在 post api 會壞掉，目前加上後已經可以順利使用！

**Comment by addie-tyc** [comment]
LGTM

**Pull request merged**